File: game/transport/browser.lua
local contract = require("game.transport.contract")

---@class BrowserTransportOptions
---@field queue_limit integer?
---@field eval fun(command: string): string? -- Test seam; defaults to love.js.eval.

---@class BrowserTransport: TransportAdapter
---@field _state TransportState
---@field _queue_limit integer
---@field _eval fun(command: string): string?
---@field _event_depth integer
---@field _dropped_outbound integer
---@field _dropped_inbound integer
---@field _malformed integer
---@field _unsupported_version integer
---@field _overflow integer
---@field _sent integer
---@field _received integer
---@field _last_error string?
local BrowserTransport = {}
BrowserTransport.__index = BrowserTransport

---@param command string
---@return string?, string?
local function default_eval(command)
    local love_api = rawget(_G, "love")
    local js = love_api and rawget(love_api, "js")
    local eval = js and rawget(js, "eval")
    if type(eval) ~= "function" then
        return nil, "love.js.eval is not available"
    end
    local ok, result = pcall(eval, command)
    if not ok then
        return nil, tostring(result)
    end
    return result == nil and "" or tostring(result)
end

---@param value string
---@return string[]
local function split_fields(value)
    local fields = {}
    local start = 1
    while true do
        local separator = value:find("|", start, true)
        if not separator then
            fields[#fields + 1] = value:sub(start)
            return fields
        end
        fields[#fields + 1] = value:sub(start, separator - 1)
        start = separator + 1
    end
end

---@param options BrowserTransportOptions?
---@return BrowserTransport
function BrowserTransport.new(options)
    options = options or {}
    local queue_limit = options.queue_limit or contract.DEFAULT_QUEUE_LIMIT
    assert(
        queue_limit == math.floor(queue_limit)
            and queue_limit > 0
            and queue_limit <= contract.MAX_QUEUE_LIMIT,
        "browser transport queue_limit is outside the supported range"
    )
    return setmetatable({
        _state = "new",
        _queue_limit = queue_limit,
        _eval = options.eval or default_eval,
        _event_depth = 0,
        _dropped_outbound = 0,
        _dropped_inbound = 0,
        _malformed = 0,
        _unsupported_version = 0,
        _overflow = 0,
        _sent = 0,
        _received = 0,
        _last_error = nil,
    }, BrowserTransport)
end

---@param code TransportErrorCode
---@param message string
function BrowserTransport:_record_error(code, message)
    self._last_error = message
    if code == "malformed" or code == "payload_too_large" then
        self._malformed = self._malformed + 1
    elseif code == "unsupported_version" then
        self._unsupported_version = self._unsupported_version + 1
    elseif code == "overflow" then
        self._overflow = self._overflow + 1
    end
end

---@param name string
---@param argument string?
---@return string?, string?
function BrowserTransport:_call(name, argument)
    local command = "window.GoliseoTransportBridge." .. name .. "("
    if argument then
        command = command .. argument
    end
    command = command .. ")"
    local result, err = self._eval(command)
    if not result then
        self:_record_error("bridge_error", err or "browser bridge returned no result")
        return nil, err or "browser bridge returned no result"
    end
    return result
end

---@param result string
---@return TransportState?, string?
local function parse_state(result)
    local fields = split_fields(result)
    if fields[1] ~= "state" or not fields[2] then
        return nil, "browser bridge returned an invalid state response"
    end
    if
        fields[2] ~= "new"
        and fields[2] ~= "connected"
        and fields[2] ~= "disconnected"
        and fields[2] ~= "closed"
        and fields[2] ~= "error"
    then
        return nil, "browser bridge returned an unknown state"
    end
    ---@cast fields[2] TransportState
    return fields[2]
end

---@param result string
---@return string?, TransportErrorCode?, string?
local function parse_result(result)
    if result == "ok" then
        return "ok"
    end
    local fields = split_fields(result)
    if fields[1] ~= "error" or not fields[2] then
        return nil, "bridge_error", "browser bridge returned an invalid operation response"
    end
    local code = fields[2]
    ---@cast code TransportErrorCode
    return nil, code, fields[3] or code
end

---@return boolean?, string?, TransportErrorCode?
function BrowserTransport:initialize()
    if self._state == "connected" then
        return true
    end
    local result, err = self:_call("initialize", tostring(self._queue_limit))
    if not result then
        return nil, err, "bridge_error"
    end
    local state, state_err = parse_state(result)
    if not state then
        local error_message = state_err or "browser bridge returned an invalid state"
        self:_record_error("bridge_error", error_message)
        return nil, error_message, "bridge_error"
    end
    self._state = state
    if state ~= "connected" then
        return nil, "browser transport did not connect", "not_connected"
    end
    return true
end

---@return boolean?, string?, TransportErrorCode?
function BrowserTransport:shutdown()
    if self._state == "closed" then
        return true
    end
    local result, err = self:_call("shutdown")
    if not result then
        return nil, err, "bridge_error"
    end
    local state, state_err = parse_state(result)
    if not state then
        local error_message = state_err or "browser bridge returned an invalid state"
        self:_record_error("bridge_error", error_message)
        return nil, error_message, "bridge_error"
    end
    self._state = state
    if state ~= "closed" then
        return nil, "browser transport did not close", "bridge_error"
    end
    return true
end

---@return boolean?, string?, TransportErrorCode?
function BrowserTransport:_require_connected()
    if self._state == "new" then
        return nil, "transport is not initialized", "not_initialized"
    end
    if self._state == "closed" then
        return nil, "transport is closed", "closed"
    end
    if self._state ~= "connected" then
        return nil, "transport is not connected", "not_connected"
    end
    return true
end

---@param message TransportMessage
---@return boolean?, string?, TransportErrorCode?
function BrowserTransport:enqueue(message)
    local ok, err, code = self:_require_connected()
    if not ok then
        return nil, err, code
    end
    ok, err, code = contract.validate(message)
    if not ok then
        local error_code = code or "malformed"
        local error_message = err or "browser transport rejected a message"
        self:_record_error(error_code, error_message)
        return nil, error_message, error_code
    end
    local wire = assert(contract.encode(message))
    local result, call_err = self:_call("enqueue", "'" .. wire .. "'")
    if not result then
        return nil, call_err, "bridge_error"
    end
    local status, result_code, result_err = parse_result(result)
    if not status then
        local error_code = result_code or "bridge_error"
        local error_message = result_err or "browser bridge rejected a message"
        self:_record_error(error_code, error_message)
        if error_code == "overflow" then
            self._dropped_outbound = self._dropped_outbound + 1
        end
        return nil, error_message, error_code
    end
    self._sent = self._sent + 1
    return true
end

---@param message TransportMessage
---@return boolean?, string?, TransportErrorCode?
function BrowserTransport:send(message)
    return self:enqueue(message)
end

---@return TransportMessage?, string?, TransportErrorCode?
function BrowserTransport:poll()
    local ok, err, code = self:_require_connected()
    if not ok then
        return nil, err, code
    end
    local wire, call_err = self:_call("poll")
    if not wire then
        return nil, call_err, "bridge_error"
    end
    if wire == "" then
        return nil
    end
    local message, decode_err, decode_code = contract.decode(wire)
    if not message then
        local error_code = decode_code or "malformed"
        local error_message = decode_err or "browser bridge returned a malformed message"
        self:_record_error(error_code, error_message)
        self._dropped_inbound = self._dropped_inbound + 1
        return nil, error_message, error_code
    end
    self._received = self._received + 1
    return message
end

---@return TransportEvent?
function BrowserTransport:poll_event()
    local result = self:_call("poll_event")
    if not result or result == "" then
        return nil
    end
    local fields = split_fields(result)
    if fields[1] == "state" and fields[2] then
        self._state = fields[2]
        return { kind = "state", state = fields[2] }
    end
    if fields[1] == "error" and fields[2] then
        self._event_depth = math.max(0, self._event_depth - 1)
        return { kind = "error", code = fields[2] }
    end
    self:_record_error("bridge_error", "browser bridge returned an invalid event response")
    return { kind = "error", code = "bridge_error", message = result }
end

---@param reason string?
---@return boolean?, string?, TransportErrorCode?
function BrowserTransport:disconnect(reason)
    local argument = reason
            and ("'" .. reason:gsub("[^%w%-%._~]", function(character)
                return ("%%%02X"):format(string.byte(character))
            end) .. "'")
        or nil
    local result, err = self:_call("disconnect", argument)
    if not result then
        return nil, err, "bridge_error"
    end
    local state, state_err = parse_state(result)
    if not state then
        local error_message = state_err or "browser bridge returned an invalid state"
        self:_record_error("bridge_error", error_message)
        return nil, error_message, "bridge_error"
    end
    self._state = state
    return true
end

---@return TransportState
function BrowserTransport:state()
    return self._state
end

---@return TransportDiagnostics
function BrowserTransport:diagnostics()
    local result, err = self:_call("diagnostics")
    if not result then
        return {
            state = self._state,
            queue_limit = self._queue_limit,
            outbound_depth = 0,
            inbound_depth = 0,
            event_depth = self._event_depth,
            dropped_outbound = self._dropped_outbound,
            dropped_inbound = self._dropped_inbound,
            malformed = self._malformed,
            unsupported_version = self._unsupported_version,
            overflow = self._overflow,
            sent = self._sent,
            received = self._received,
            last_error = err,
        }
    end
    local fields = split_fields(result)
    if #fields ~= 13 then
        self:_record_error("bridge_error", "browser bridge returned invalid diagnostics")
        return self:diagnostics_fallback()
    end
    local numbers = {}
    for index = 2, 12 do
        numbers[index] = tonumber(fields[index])
        if not numbers[index] then
            self:_record_error("bridge_error", "browser bridge diagnostics contain invalid numbers")
            return self:diagnostics_fallback()
        end
    end
    self._state = fields[1]
    return {
        state = fields[1],
        queue_limit = numbers[2],
        outbound_depth = numbers[3],
        inbound_depth = numbers[4],
        event_depth = numbers[5],
        dropped_outbound = numbers[6],
        dropped_inbound = numbers[7],
        malformed = numbers[8],
        unsupported_version = numbers[9],
        overflow = numbers[10],
        sent = numbers[11],
        received = numbers[12],
        last_error = fields[13] ~= "" and fields[13] or nil,
    }
end

---@return TransportDiagnostics
function BrowserTransport:diagnostics_fallback()
    return {
        state = self._state,
        queue_limit = self._queue_limit,
        outbound_depth = 0,
        inbound_depth = 0,
        event_depth = self._event_depth,
        dropped_outbound = self._dropped_outbound,
        dropped_inbound = self._dropped_inbound,
        malformed = self._malformed,
        unsupported_version = self._unsupported_version,
        overflow = self._overflow,
        sent = self._sent,
        received = self._received,
        last_error = self._last_error,
    }
end

return BrowserTransport

File: ts/packages/transport/src/browser.ts
// Real browser transport, driven through a single bounded-string eval seam
// against the DOM `window` bridge.

import { ok, err } from "@gc/core";
import * as contract from "./contract.ts";
import type {
  TransportAdapter,
  TransportDiagnostics,
  TransportErrorCode,
  TransportEvent,
  TransportMessage,
  TransportResult,
  TransportState,
} from "./contract.ts";

/**
 * Test seam; defaults to evaluating `window.GoliseoTransportBridge`.
 * Returns a `[result, error]` pair so a mock is free to explain a `null`
 * result with a message.
 */
export type EvalFn = (command: string) => readonly [result: string | null, error: string | null];

export interface BrowserTransportOptions {
  readonly queue_limit?: number;
  readonly eval?: EvalFn;
}

function isBrowserState(value: string): value is TransportState {
  return (
    value === "new" ||
    value === "connected" ||
    value === "disconnected" ||
    value === "closed" ||
    value === "error"
  );
}

function splitFields(value: string): string[] {
  const fields: string[] = [];
  let start = 0;
  for (;;) {
    const separator = value.indexOf("|", start);
    if (separator === -1) {
      fields.push(value.slice(start));
      return fields;
    }
    fields.push(value.slice(start, separator));
    start = separator + 1;
  }
}

function parseState(result: string): { readonly state: TransportState | null; readonly error?: string } {
  const fields = splitFields(result);
  const tag = fields[0];
  const rawState = fields[1];
  if (tag !== "state" || rawState === undefined) {
    return { state: null, error: "browser bridge returned an invalid state response" };
  }
  if (!isBrowserState(rawState)) {
    return { state: null, error: "browser bridge returned an unknown state" };
  }
  return { state: rawState };
}

interface ParsedResult {
  readonly ok: boolean;
  readonly code?: TransportErrorCode;
  readonly detail?: string;
}

function parseResult(result: string): ParsedResult {
  if (result === "ok") {
    return { ok: true };
  }
  const fields = splitFields(result);
  const tag = fields[0];
  const code = fields[1];
  if (tag !== "error" || code === undefined) {
    return { ok: false, code: "bridge_error", detail: "browser bridge returned an invalid operation response" };
  }
  return { ok: false, code: code as TransportErrorCode, detail: fields[2] ?? code };
}

/** Real browser transport: real WebRTC/WebSocket behind a bounded eval seam. */
export class BrowserTransport implements TransportAdapter {
  private _state: TransportState = "new";
  private readonly _queueLimit: number;
  private readonly _eval: EvalFn;
  private _eventDepth = 0;
  private _droppedOutbound = 0;
  private _droppedInbound = 0;
  private _malformed = 0;
  private _unsupportedVersion = 0;
  private _overflow = 0;
  private _sent = 0;
  private _received = 0;
  private _lastError: string | null = null;

  constructor(options: BrowserTransportOptions = {}) {
    const queueLimit = options.queue_limit ?? contract.DEFAULT_QUEUE_LIMIT;
    if (
      queueLimit !== Math.floor(queueLimit) ||
      queueLimit <= 0 ||
      queueLimit > contract.MAX_QUEUE_LIMIT
    ) {
      throw new Error("browser transport queue_limit is outside the supported range");
    }
    this._queueLimit = queueLimit;
    this._eval = options.eval ?? BrowserTransport._defaultEval;
  }

  private static _defaultEval(_command: string): readonly [result: string | null, error: string | null] {
    // The real DOM bridge (`window.GoliseoTransportBridge`) is wired up by
    // the app shell at runtime; this package implements transport behavior
    // only, not those bindings, so the default seam has nothing to call.
    return [null, "window.GoliseoTransportBridge is not available"];
  }

  private _recordError(code: TransportErrorCode, message: string): void {
    this._lastError = message;
    if (code === "malformed" || code === "payload_too_large") {
      this._malformed += 1;
    } else if (code === "unsupported_version") {
      this._unsupportedVersion += 1;
    } else if (code === "overflow") {
      this._overflow += 1;
    }
  }

  private _call(name: string, argument?: string): { readonly result: string | null; readonly error?: string } {
    const command = `window.GoliseoTransportBridge.${name}(${argument ?? ""})`;
    const [result, evalError] = this._eval(command);
    if (result === null) {
      const message = evalError ?? "browser bridge returned no result";
      this._recordError("bridge_error", message);
      return { result: null, error: message };
    }
    return { result };
  }

  initialize(): TransportResult<true> {
    if (this._state === "connected") {
      return ok(true);
    }
    const { result, error } = this._call("initialize", String(this._queueLimit));
    if (result === null) {
      return err({ message: error ?? "browser bridge returned no result", code: "bridge_error" });
    }
    const { state, error: stateError } = parseState(result);
    if (state === null) {
      const message = stateError ?? "browser bridge returned an invalid state";
      this._recordError("bridge_error", message);
      return err({ message, code: "bridge_error" });
    }
    this._state = state;
    if (state !== "connected") {
      return err({ message: "browser transport did not connect", code: "not_connected" });
    }
    return ok(true);
  }

  shutdown(): TransportResult<true> {
    if (this._state === "closed") {
      return ok(true);
    }
    const { result, error } = this._call("shutdown");
    if (result === null) {
      return err({ message: error ?? "browser bridge returned no result", code: "bridge_error" });
    }
    const { state, error: stateError } = parseState(result);
    if (state === null) {
      const message = stateError ?? "browser bridge returned an invalid state";
      this._recordError("bridge_error", message);
      return err({ message, code: "bridge_error" });
    }
    this._state = state;
    if (state !== "closed") {
      return err({ message: "browser transport did not close", code: "bridge_error" });
    }
    return ok(true);
  }

  private _requireConnected(): TransportResult<true> {
    if (this._state === "new") {
      return err({ message: "transport is not initialized", code: "not_initialized" });
    }
    if (this._state === "closed") {
      return err({ message: "transport is closed", code: "closed" });
    }
    if (this._state !== "connected") {
      return err({ message: "transport is not connected", code: "not_connected" });
    }
    return ok(true);
  }

  enqueue(message: TransportMessage): TransportResult<true> {
    const connected = this._requireConnected();
    if (!connected.ok) {
      return connected;
    }
    const validated = contract.validate(message);
    if (!validated.ok) {
      this._recordError(validated.error.code, validated.error.message);
      return validated;
    }
    const wireResult = contract.encode(message);
    if (!wireResult.ok) {
      // `validate` above already guarantees `encode` succeeds.
      throw new Error(wireResult.error.message);
    }
    const { result, error: callError } = this._call("enqueue", `'${wireResult.value}'`);
    if (result === null) {
      return err({ message: callError ?? "browser bridge returned no result", code: "bridge_error" });
    }
    const parsed = parseResult(result);
    if (!parsed.ok) {
      const code = parsed.code ?? "bridge_error";
      const message = parsed.detail ?? "browser bridge rejected a message";
      this._recordError(code, message);
      if (code === "overflow") {
        this._droppedOutbound += 1;
      }
      return err({ message, code });
    }
    this._sent += 1;
    return ok(true);
  }

  send(message: TransportMessage): TransportResult<true> {
    return this.enqueue(message);
  }

  poll(): TransportResult<TransportMessage | null> {
    const connected = this._requireConnected();
    if (!connected.ok) {
      return connected;
    }
    const { result, error: callError } = this._call("poll");
    if (result === null) {
      return err({ message: callError ?? "browser bridge returned no result", code: "bridge_error" });
    }
    if (result === "") {
      return ok(null);
    }
    const decoded = contract.decode(result);
    if (!decoded.ok) {
      this._recordError(decoded.error.code, decoded.error.message);
      this._droppedInbound += 1;
      return decoded;
    }
    this._received += 1;
    return ok(decoded.value);
  }

  pollEvent(): TransportEvent | null {
    const { result } = this._call("poll_event");
    if (result === null || result === "") {
      return null;
    }
    const fields = splitFields(result);
    const tag = fields[0];
    if (tag === "state" && fields[1] !== undefined && fields[1] !== "") {
      const state = fields[1];
      if (isBrowserState(state)) {
        this._state = state;
        return { kind: "state", state };
      }
    }
    if (tag === "error" && fields[1] !== undefined && fields[1] !== "") {
      this._eventDepth = Math.max(0, this._eventDepth - 1);
      return { kind: "error", code: fields[1] as TransportErrorCode };
    }
    this._recordError("bridge_error", "browser bridge returned an invalid event response");
    return { kind: "error", code: "bridge_error", message: result };
  }

  disconnect(reason?: string): TransportResult<true> {
    const argument =
      reason !== undefined
        ? `'${reason.replace(/[^A-Za-z0-9\-._~]/g, (character) => "%" + (character.charCodeAt(0) & 0xff).toString(16).toUpperCase().padStart(2, "0"))}'`
        : undefined;
    const { result, error } = this._call("disconnect", argument);
    if (result === null) {
      return err({ message: error ?? "browser bridge returned no result", code: "bridge_error" });
    }
    const { state, error: stateError } = parseState(result);
    if (state === null) {
      const message = stateError ?? "browser bridge returned an invalid state";
      this._recordError("bridge_error", message);
      return err({ message, code: "bridge_error" });
    }
    this._state = state;
    return ok(true);
  }

  state(): TransportState {
    return this._state;
  }

  diagnostics(): TransportDiagnostics {
    const { result, error } = this._call("diagnostics");
    if (result === null) {
      return this._diagnosticsFallback(error ?? null);
    }
    const fields = splitFields(result);
    if (fields.length !== 13) {
      this._recordError("bridge_error", "browser bridge returned invalid diagnostics");
      return this._diagnosticsFallback(null);
    }
    const numbers: number[] = [];
    for (let index = 1; index <= 11; index += 1) {
      const raw = fields[index];
      const value = raw === undefined ? NaN : Number(raw);
      if (!Number.isFinite(value)) {
        this._recordError("bridge_error", "browser bridge diagnostics contain invalid numbers");
        return this._diagnosticsFallback(null);
      }
      numbers[index] = value;
    }
    const stateField = fields[0];
    if (stateField === undefined || !isBrowserState(stateField)) {
      this._recordError("bridge_error", "browser bridge returned invalid diagnostics");
      return this._diagnosticsFallback(null);
    }
    this._state = stateField;
    const lastErrorField = fields[12];
    return {
      state: stateField,
      queue_limit: numbers[1] as number,
      outbound_depth: numbers[2] as number,
      inbound_depth: numbers[3] as number,
      event_depth: numbers[4] as number,
      dropped_outbound: numbers[5] as number,
      dropped_inbound: numbers[6] as number,
      malformed: numbers[7] as number,
      unsupported_version: numbers[8] as number,
      overflow: numbers[9] as number,
      sent: numbers[10] as number,
      received: numbers[11] as number,
      last_error: lastErrorField !== undefined && lastErrorField !== "" ? lastErrorField : null,
    };
  }

  private _diagnosticsFallback(error: string | null): TransportDiagnostics {
    return {
      state: this._state,
      queue_limit: this._queueLimit,
      outbound_depth: 0,
      inbound_depth: 0,
      event_depth: this._eventDepth,
      dropped_outbound: this._droppedOutbound,
      dropped_inbound: this._droppedInbound,
      malformed: this._malformed,
      unsupported_version: this._unsupportedVersion,
      overflow: this._overflow,
      sent: this._sent,
      received: this._received,
      last_error: error ?? this._lastError,
    };
  }
}
